File: test/example/native/android/app/src/main/java/dev/sigra/proof/HostedAuthSession.kt
package dev.sigra.proof

import android.app.Activity
import android.content.Context
import android.net.Uri
import androidx.activity.result.ActivityResultLauncher
import androidx.browser.auth.AuthTabIntent
import androidx.browser.customtabs.CustomTabsClient
import androidx.browser.customtabs.CustomTabsIntent
import java.io.BufferedReader
import java.net.HttpURLConnection
import java.net.URI
import java.net.URL
import java.net.URLDecoder
import java.net.URLEncoder
import java.nio.charset.StandardCharsets
import java.security.MessageDigest
import java.security.SecureRandom
import java.util.Base64

data class HostedAuthConfiguration(val baseUrl: String, val profileId: String)

data class HostedAuthorizationAttempt(
    val state: String,
    val verifier: String,
    val startUri: URI,
    val query: Map<String, String>,
)

data class HostedSessionMaterial(val access: ByteArray, val refresh: ByteArray)

enum class BrowserMode(val wireValue: String) {
    AUTH_TAB("auth_tab"),
    CUSTOM_TAB_FALLBACK("custom_tab_fallback");

    companion object {
        fun fromWire(value: String): BrowserMode = entries.singleOrNull { it.wireValue == value }
            ?: throw IllegalArgumentException("browser mode is not allowlisted")
    }
}

sealed class HostedAuthException(message: String) : IllegalArgumentException(message) {
    class InvalidConfiguration : HostedAuthException("invalid hosted auth configuration")
    class InvalidCallback : HostedAuthException("invalid hosted auth callback")
    class InvalidResponse : HostedAuthException("invalid hosted auth response")
    class ServerRevoked : HostedAuthException("host revoked the session")
}

interface HostedAuthTransport {
    fun exchange(code: String, verifier: String, configuration: HostedAuthConfiguration): HostedSessionMaterial
    fun refresh(material: ByteArray, configuration: HostedAuthConfiguration): HostedSessionMaterial
}

class HttpHostedAuthTransport : HostedAuthTransport {
    override fun exchange(
        code: String,
        verifier: String,
        configuration: HostedAuthConfiguration,
    ): HostedSessionMaterial = request(
        "/api/app-login/exchange",
        mapOf(
            "code" to code,
            "code_verifier" to verifier,
            "profile_id" to configuration.profileId,
            "callback" to HostedAuthSession.CALLBACK_URI,
        ),
        configuration,
    )

    override fun refresh(
        material: ByteArray,
        configuration: HostedAuthConfiguration,
    ): HostedSessionMaterial = request(
        "/api/app-login/refresh",
        mapOf("refresh_token" to material.toString(StandardCharsets.UTF_8)),
        configuration,
    )

    private fun request(
        path: String,
        body: Map<String, String>,
        configuration: HostedAuthConfiguration,
    ): HostedSessionMaterial {
        val base = URI(configuration.baseUrl)
        if (base.scheme !in setOf("http", "https") || base.host.isNullOrBlank()) {
            throw HostedAuthException.InvalidConfiguration()
        }
        val connection = URL(configuration.baseUrl.trimEnd('/') + path).openConnection() as HttpURLConnection
        connection.requestMethod = "POST"
        connection.connectTimeout = 10_000
        connection.readTimeout = 10_000
        connection.setRequestProperty("Content-Type", "application/json")
        connection.doOutput = true
        connection.outputStream.use { output -> output.write(jsonObject(body).toByteArray(StandardCharsets.UTF_8)) }
        if (connection.responseCode == HttpURLConnection.HTTP_UNAUTHORIZED) {
            throw HostedAuthException.ServerRevoked()
        }
        if (connection.responseCode !in 200..299) throw HostedAuthException.InvalidResponse()
        val response = connection.inputStream.bufferedReader().use(BufferedReader::readText)
        val access = jsonString(response, "access_token")?.toByteArray(StandardCharsets.UTF_8)
        val refresh = jsonString(response, "refresh_token")?.toByteArray(StandardCharsets.UTF_8)
        if (access == null || access.isEmpty() || refresh == null || refresh.isEmpty()) {
            throw HostedAuthException.InvalidResponse()
        }
        return HostedSessionMaterial(access, refresh)
    }

    private fun jsonObject(fields: Map<String, String>): String = fields.entries.joinToString(
        prefix = "{",
        postfix = "}",
    ) { (key, value) -> "\"${escape(key)}\":\"${escape(value)}\"" }

    private fun jsonString(value: String, key: String): String? {
        val pattern = Regex("\\\"${Regex.escape(key)}\\\"\\s*:\\s*\\\"((?:\\\\.|[^\\\"\\\\])*)\\\"")
        return pattern.find(value)?.groupValues?.get(1)?.let(::unescape)
    }

    private fun escape(value: String): String = value
        .replace("\\", "\\\\")
        .replace("\"", "\\\"")
        .replace("\n", "\\n")

    private fun unescape(value: String): String = value
        .replace("\\n", "\n")
        .replace("\\\"", "\"")
        .replace("\\\\", "\\")
}

class HostedAuthSession(
    private val configuration: HostedAuthConfiguration,
    private val transport: HostedAuthTransport,
    private val refreshStore: SecureRefreshStore,
) {
    private var accessMaterial: ByteArray? = null

    val hasMemoryOnlyAccess: Boolean get() = accessMaterial != null
    val storagePosture: StoragePosture get() = refreshStore.posture

    fun completeCallback(callback: String, expectedState: String, verifier: String): StoragePosture {
        val code = validatedCode(callback, expectedState)
        val issued = transport.exchange(code, verifier, configuration)
        accessMaterial = issued.access.copyOf()
        return refreshStore.saveInitial(issued.refresh)
    }

    fun rotateRefresh(): StoragePosture = try {
        val replacement = transport.refresh(refreshStore.currentMaterial(), configuration)
        accessMaterial = replacement.access.copyOf()
        refreshStore.rotate(replacement.refresh)
    } catch (_: HostedAuthException.ServerRevoked) {
        accessMaterial = null
        refreshStore.deleteAfterRevocation()
    }

    fun recoverAfterRelaunch(): StoragePosture {
        accessMaterial = null
        return refreshStore.recoverAfterRelaunch()
    }

    internal fun <T> withMemoryOnlyAccess(block: (ByteArray) -> T): T {
        val material = accessMaterial ?: throw HostedAuthException.InvalidResponse()
        return block(material.copyOf())
    }

    fun logout(): StoragePosture {
        accessMaterial = null
        return refreshStore.deleteAfterLogout()
    }

    fun markServerRevoked(): StoragePosture {
        accessMaterial = null
        return refreshStore.deleteAfterRevocation()
    }

    companion object {
        const val CALLBACK_URI = "sigra-native-proof://auth/android"
        const val LOCKED_BROWSER_PACKAGE = "com.android.chrome"

        fun authorizationRequest(
            configuration: HostedAuthConfiguration,
            state: String = randomUrlSafe(32),
            verifier: String = randomUrlSafe(48),
        ): HostedAuthorizationAttempt {
            val base = runCatching { URI(configuration.baseUrl) }.getOrNull()
            if (base == null || base.scheme !in setOf("http", "https") || base.host.isNullOrBlank() ||
                configuration.profileId.isBlank() || state.isBlank() || verifier.length !in 43..128 ||
                !verifier.matches(Regex("[A-Za-z0-9._~-]+"))
            ) throw HostedAuthException.InvalidConfiguration()
            val query = linkedMapOf(
                "profile_id" to configuration.profileId,
                "callback" to CALLBACK_URI,
                "state" to state,
                "code_challenge" to pkceChallenge(verifier),
                "code_challenge_method" to "S256",
            )
            val encoded = query.entries.joinToString("&") { (key, value) ->
                "${urlEncode(key)}=${urlEncode(value)}"
            }
            val uri = URI(base.scheme, base.userInfo, base.host, base.port, "/users/app-login", encoded, null)
            return HostedAuthorizationAttempt(state, verifier, uri, query)
        }

        fun validatedCode(callback: String, expectedState: String): String {
            val uri = runCatching { URI(callback) }.getOrNull() ?: throw HostedAuthException.InvalidCallback()
            if (uri.scheme != "sigra-native-proof" || uri.host != "auth" || uri.path != "/android" ||
                uri.port != -1 || uri.userInfo != null || uri.fragment != null
            ) throw HostedAuthException.InvalidCallback()
            val items = parseQuery(uri.rawQuery)
            if (items.keys != setOf("code", "state") || items["code"].isNullOrBlank() ||
                !constantTimeEqual(items["state"].orEmpty(), expectedState)
            ) throw HostedAuthException.InvalidCallback()
            return items.getValue("code")
        }

        fun selectBrowserMode(lockedMode: BrowserMode, providerSupportsAuthTab: Boolean): BrowserMode =
            if (lockedMode == BrowserMode.AUTH_TAB && providerSupportsAuthTab) BrowserMode.AUTH_TAB
            else BrowserMode.CUSTOM_TAB_FALLBACK

        fun shouldLaunchAuthTab(mode: BrowserMode): Boolean = mode == BrowserMode.AUTH_TAB

        fun launchBrowser(
            activity: Activity,
            authTabLauncher: ActivityResultLauncher<android.content.Intent>,
            startUri: URI,
            lockedMode: BrowserMode,
        ): BrowserMode {
            val providerAvailable = activity.packageManager.getLaunchIntentForPackage(LOCKED_BROWSER_PACKAGE) != null
            if (!providerAvailable) throw HostedAuthException.InvalidConfiguration()
            val supportsAuthTab = CustomTabsClient.isAuthTabSupported(activity, LOCKED_BROWSER_PACKAGE)
            val selected = selectBrowserMode(lockedMode, supportsAuthTab)
            if (selected == BrowserMode.AUTH_TAB) {
                AuthTabIntent.Builder().build().launch(authTabLauncher, Uri.parse(startUri.toString()), "sigra-native-proof")
            } else {
                CustomTabsIntent.Builder().build().also { it.intent.setPackage(LOCKED_BROWSER_PACKAGE) }
                    .launchUrl(activity, Uri.parse(startUri.toString()))
            }
            return selected
        }

        fun providerSupportsAuthTab(context: Context): Boolean =
            CustomTabsClient.isAuthTabSupported(context, LOCKED_BROWSER_PACKAGE)

        private fun parseQuery(raw: String?): Map<String, String> {
            if (raw.isNullOrEmpty()) return emptyMap()
            val result = linkedMapOf<String, String>()
            raw.split('&').forEach { pair ->
                val pieces = pair.split('=', limit = 2)
                if (pieces.size != 2) throw HostedAuthException.InvalidCallback()
                val key = URLDecoder.decode(pieces[0], StandardCharsets.UTF_8.name())
                val value = URLDecoder.decode(pieces[1], StandardCharsets.UTF_8.name())
                if (result.put(key, value) != null) throw HostedAuthException.InvalidCallback()
            }
            return result
        }

        private fun pkceChallenge(verifier: String): String = Base64.getUrlEncoder().withoutPadding()
            .encodeToString(MessageDigest.getInstance("SHA-256").digest(verifier.toByteArray(StandardCharsets.US_ASCII)))

        private fun randomUrlSafe(byteCount: Int): String = ByteArray(byteCount).also(SecureRandom()::nextBytes)
            .let { Base64.getUrlEncoder().withoutPadding().encodeToString(it) }

        private fun urlEncode(value: String): String =
            URLEncoder.encode(value, StandardCharsets.UTF_8.name()).replace("+", "%20")

        private fun constantTimeEqual(left: String, right: String): Boolean = MessageDigest.isEqual(
            left.toByteArray(StandardCharsets.UTF_8),
            right.toByteArray(StandardCharsets.UTF_8),
        )
    }
}

File: test/example/native/ios/SigraNativeProof/App/HostedAuthSession.swift
import AuthenticationServices
import CryptoKit
import Foundation
import Security

struct HostedAuthConfiguration: Equatable {
    let baseURL: URL
    let profileID: String
}

struct HostedAuthorizationAttempt: Equatable {
    let state: String
    let verifier: String
    let startURL: URL
}

struct HostedSessionMaterial: Equatable {
    let access: Data
    let refresh: Data
}

enum HostedAuthError: Error, Equatable {
    case invalidConfiguration
    case invalidCallback
    case browserCancelled
    case invalidResponse
    case serverRevoked
    case transportFailure
}

protocol HostedAuthTransport {
    func exchange(
        code: String,
        verifier: String,
        configuration: HostedAuthConfiguration
    ) async throws -> HostedSessionMaterial
    func refresh(_ material: Data, configuration: HostedAuthConfiguration) async throws -> HostedSessionMaterial
}

struct URLSessionHostedAuthTransport: HostedAuthTransport {
    private struct SessionResponse: Decodable {
        let accessToken: String
        let refreshToken: String

        enum CodingKeys: String, CodingKey {
            case accessToken = "access_token"
            case refreshToken = "refresh_token"
        }
    }

    private let session: URLSession

    init(session: URLSession = .shared) {
        self.session = session
    }

    func exchange(
        code: String,
        verifier: String,
        configuration: HostedAuthConfiguration
    ) async throws -> HostedSessionMaterial {
        let body: [String: String] = [
            "code": code,
            "code_verifier": verifier,
            "profile_id": configuration.profileID,
            "callback": HostedAuthSession.callbackURL.absoluteString
        ]
        return try await request(path: "/api/app-login/exchange", body: body, configuration: configuration)
    }

    func refresh(_ material: Data, configuration: HostedAuthConfiguration) async throws -> HostedSessionMaterial {
        guard let value = String(data: material, encoding: .utf8), !value.isEmpty else {
            throw HostedAuthError.invalidResponse
        }
        return try await request(
            path: "/api/app-login/refresh",
            body: ["refresh_token": value],
            configuration: configuration
        )
    }

    private func request(
        path: String,
        body: [String: String],
        configuration: HostedAuthConfiguration
    ) async throws -> HostedSessionMaterial {
        guard var components = URLComponents(url: configuration.baseURL, resolvingAgainstBaseURL: false),
              components.scheme == "https" || components.scheme == "http",
              components.host != nil else {
            throw HostedAuthError.invalidConfiguration
        }
        components.path = path
        components.query = nil
        guard let url = components.url else { throw HostedAuthError.invalidConfiguration }
        var request = URLRequest(url: url)
        request.httpMethod = "POST"
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        request.httpBody = try JSONEncoder().encode(body)

        let (data, response): (Data, URLResponse)
        do {
            (data, response) = try await session.data(for: request)
        } catch {
            throw HostedAuthError.transportFailure
        }
        guard let http = response as? HTTPURLResponse else { throw HostedAuthError.invalidResponse }
        if http.statusCode == 401 { throw HostedAuthError.serverRevoked }
        guard (200..<300).contains(http.statusCode) else { throw HostedAuthError.invalidResponse }
        let decoded = try JSONDecoder().decode(SessionResponse.self, from: data)
        guard let access = decoded.accessToken.data(using: .utf8), !access.isEmpty,
              let refresh = decoded.refreshToken.data(using: .utf8), !refresh.isEmpty else {
            throw HostedAuthError.invalidResponse
        }
        return HostedSessionMaterial(access: access, refresh: refresh)
    }
}

@MainActor
final class HostedAuthSession: NSObject, ASWebAuthenticationPresentationContextProviding {
    nonisolated static let callbackURL = URL(string: "sigra-native-proof://auth/callback")!

    private let configuration: HostedAuthConfiguration
    private let transport: HostedAuthTransport
    private let refreshStore: SecureRefreshStore
    private var accessMaterial: Data?
    private var webSession: ASWebAuthenticationSession?
    private var presentationAnchor: ASPresentationAnchor?

    var hasMemoryOnlyAccess: Bool { accessMaterial != nil }
    var storagePosture: StoragePosture { refreshStore.posture }

    init(
        configuration: HostedAuthConfiguration,
        transport: HostedAuthTransport = URLSessionHostedAuthTransport(),
        refreshStore: SecureRefreshStore = SecureRefreshStore()
    ) {
        self.configuration = configuration
        self.transport = transport
        self.refreshStore = refreshStore
    }

    nonisolated static func authorizationRequest(
        configuration: HostedAuthConfiguration,
        state: String,
        verifier: String
    ) throws -> HostedAuthorizationAttempt {
        guard !state.isEmpty,
              (43...128).contains(verifier.utf8.count),
              verifier.unicodeScalars.allSatisfy({
                  CharacterSet(charactersIn: "abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789-._~")
                      .contains($0)
              }),
              var components = URLComponents(url: configuration.baseURL, resolvingAgainstBaseURL: false),
              components.scheme == "https" || components.scheme == "http",
              components.host != nil,
              !configuration.profileID.isEmpty else {
            throw HostedAuthError.invalidConfiguration
        }
        components.path = "/users/app-login"
        components.queryItems = [
            URLQueryItem(name: "profile_id", value: configuration.profileID),
            URLQueryItem(name: "callback", value: callbackURL.absoluteString),
            URLQueryItem(name: "state", value: state),
            URLQueryItem(name: "code_challenge", value: pkceChallenge(verifier)),
            URLQueryItem(name: "code_challenge_method", value: "S256")
        ]
        guard let startURL = components.url else { throw HostedAuthError.invalidConfiguration }
        return HostedAuthorizationAttempt(state: state, verifier: verifier, startURL: startURL)
    }

    nonisolated static func validatedCode(from callback: URL, expectedState: String) throws -> String {
        guard let components = URLComponents(url: callback, resolvingAgainstBaseURL: false),
              components.scheme == callbackURL.scheme,
              components.host == callbackURL.host,
              components.path == callbackURL.path,
              components.port == nil,
              components.user == nil,
              components.password == nil,
              components.fragment == nil,
              let items = components.queryItems,
              items.map(\URLQueryItem.name).sorted() == ["code", "state"],
              let code = items.first(where: { $0.name == "code" })?.value,
              !code.isEmpty,
              let state = items.first(where: { $0.name == "state" })?.value,
              constantTimeEqual(state, expectedState) else {
            throw HostedAuthError.invalidCallback
        }
        return code
    }

    func startHostedLogin(from anchor: ASPresentationAnchor) async throws -> StoragePosture {
        let attempt = try Self.authorizationRequest(
            configuration: configuration,
            state: try Self.randomURLSafe(byteCount: 32),
            verifier: try Self.randomURLSafe(byteCount: 48)
        )
        presentationAnchor = anchor
        let callback = try await browserCallback(for: attempt.startURL)
        let code = try Self.validatedCode(from: callback, expectedState: attempt.state)
        let issued = try await transport.exchange(
            code: code,
            verifier: attempt.verifier,
            configuration: configuration
        )
        accessMaterial = issued.access
        return try refreshStore.saveInitial(issued.refresh)
    }

    func rotateRefresh() async throws -> StoragePosture {
        let current = try refreshStore.currentMaterial()
        do {
            let replacement = try await transport.refresh(current, configuration: configuration)
            accessMaterial = replacement.access
            return try refreshStore.rotate(to: replacement.refresh)
        } catch HostedAuthError.serverRevoked {
            accessMaterial = nil
            return refreshStore.deleteAfterRevocation()
        }
    }

    func recoverAfterRelaunch() -> StoragePosture {
        accessMaterial = nil
        return refreshStore.recoverAfterRelaunch()
    }

    func logout() -> StoragePosture {
        accessMaterial = nil
        return refreshStore.deleteAfterLogout()
    }

    func markServerRevoked() -> StoragePosture {
        accessMaterial = nil
        return refreshStore.deleteAfterRevocation()
    }

    func presentationAnchor(for session: ASWebAuthenticationSession) -> ASPresentationAnchor {
        presentationAnchor ?? ASPresentationAnchor()
    }

    private func browserCallback(for startURL: URL) async throws -> URL {
        try await withCheckedThrowingContinuation { continuation in
            let session = ASWebAuthenticationSession(
                url: startURL,
                callbackURLScheme: Self.callbackURL.scheme
            ) { callback, error in
                if let callback {
                    continuation.resume(returning: callback)
                } else if error != nil {
                    continuation.resume(throwing: HostedAuthError.browserCancelled)
                } else {
                    continuation.resume(throwing: HostedAuthError.invalidCallback)
                }
            }
            session.presentationContextProvider = self
            session.prefersEphemeralWebBrowserSession = false
            webSession = session
            guard session.start() else {
                continuation.resume(throwing: HostedAuthError.transportFailure)
                return
            }
        }
    }

    nonisolated private static func pkceChallenge(_ verifier: String) -> String {
        Data(SHA256.hash(data: Data(verifier.utf8)))
            .base64EncodedString()
            .replacingOccurrences(of: "+", with: "-")
            .replacingOccurrences(of: "/", with: "_")
            .replacingOccurrences(of: "=", with: "")
    }

    nonisolated private static func randomURLSafe(byteCount: Int) throws -> String {
        var data = Data(count: byteCount)
        let result = data.withUnsafeMutableBytes { raw in
            SecRandomCopyBytes(kSecRandomDefault, byteCount, raw.baseAddress!)
        }
        guard result == errSecSuccess else { throw HostedAuthError.transportFailure }
        return data.base64EncodedString()
            .replacingOccurrences(of: "+", with: "-")
            .replacingOccurrences(of: "/", with: "_")
            .replacingOccurrences(of: "=", with: "")
    }

    nonisolated private static func constantTimeEqual(_ lhs: String, _ rhs: String) -> Bool {
        let left = Array(lhs.utf8)
        let right = Array(rhs.utf8)
        guard left.count == right.count else { return false }
        return zip(left, right).reduce(UInt8(0)) { partial, pair in partial | (pair.0 ^ pair.1) } == 0
    }
}
